### ISSUE TITLE
Build fails with "Maximum call stack size exceeded" inside the TS compiler

### ISSUE THREAD

**Issue by justingrant** [aws/aws-sdk-js-v3#1460]
**Describe the bug**
The current master branch of aws-sdk-js-v3 is not buildable on my Mac. The build fails in the client-ec2 package with a `Maximum call stack size exceeded` error inside the TS compiler. Just to make sure this wasn't an issue with the latest TS 4.0 release (only 8 days old), I tried TS 3.9.7 and got the same failure. 

I assume it's not failing in CI because the last PR wouldn't have been merged, so perhaps this is only happening for fresh local git clones?

**SDK version number**
`1.0.0-gamma.6`

**Is the issue in the browser/Node.js/ReactNative?**
Browser/Node.js/ReactNative

**Details of the browser/Node.js/ReactNative version**
node: `v14.4.0`
yarn: `v1.22.4`
macos: `10.15.6`

**To Reproduce (observed behavior)**
```
git clone https://github.com/aws/aws-sdk-js-v3.git
cd aws-sdk-js-v3
yarn
yarn build:all
```

**Actual Behavior**

```
erna ERR! yarn run build exited 1 in '@aws-sdk/client-ec2'
lerna ERR! yarn run build stdout:
$ yarn build:cjs && yarn build:es
$ tsc -p tsconfig.json
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.

lerna ERR! yarn run build stderr:
/Users/justingrant/Documents/hdev/amplify-js/aws-sdk-js-v3/node_modules/typescript/lib/tsc.js:84772
                throw e;
                ^

RangeError: Maximum call stack size exceeded
    at pipelineEmitWithHint (/Users/justingrant/Documents/hdev/amplify-js/aws-sdk-js-v3/node_modules/typescript/lib/tsc.js:80132:32)
    at pipelineEmitWithSourceMap (/Users/justingrant/Documents/hdev/amplify-js/aws-sdk-js-v3/node_modules/typescript/lib/tsc.js:83062:21)
    at pipelineEmitWithComments (/Users/justingrant/Documents/hdev/amplify-js/aws-sdk-js-v3/node_modules/typescript/lib/tsc.js:82829:17)
    at maybePipelineEmitExpression (/Users/justingrant/Documents/hdev/amplify-js/aws-sdk-js-v3/node_modules/typescript/lib/tsc.js:80850:21)
    at emitBinaryExpression (/Users/justingrant/Documents/hdev/amplify-js/aws-sdk-js-v3/node_modules/typescript/lib/tsc.js:80824:25)
    at pipelineEmitWithHint (/Users/justingrant/Documents/hdev/amplify-js/aws-sdk-js-v3/node_modules/typescript/lib/tsc.js:80140:32)
    at pipelineEmitWithSourceMap (/Users/justingrant/Documents/hdev/amplify-js/aws-sdk-js-v3/node_modules/typescript/lib/tsc.js:83062:21)
    at pipelineEmitWithComments (/Users/justingrant/Documents/hdev/amplify-js/aws-sdk-js-v3/node_modules/typescript/lib/tsc.js:82829:17)
    at maybePipelineEmitExpression (/Users/justingrant/Documents/hdev/amplify-js/aws-sdk-js-v3/node_modules/typescript/lib/tsc.js:80850:21)
    at emitBinaryExpression (/Users/justingrant/Documents/hdev/amplify-js/aws-sdk-js-v3/node_modules/typescript/lib/tsc.js:80824:25)
error Command failed with exit code 1.
error Command failed with exit code 1.

lerna ERR! yarn run build exited 1 in '@aws-sdk/client-ec2'
```

**Expected behavior**
builds successfully

**Screenshots**
n/a

**Additional context**
n/a

**Comment by trivikr** [comment]
This is being worked on in an internal issue in JS-2017

Findings by @alexforsyth:

**Background**: As part of typescript compilation Typescript attempts to construct an abstract syntax tree (AST) to represent all the types, functions, etc in your code. It uses this AST to verify that all the types line up and do the 'type' part of typescript. The way this works is a tree is set up and then typescript walks this from the root to each of the nodes in order to surface the full stack trace if there are type errors. It first does this inside each file to verify internal type consistency inside a file and then for each exported type it keeps a reference to its node in the heap's memory.

**Root cause**: Type checking inside a file is done in-memory. This is normally fine as it makes things much much faster. However as a file gets really really large this starts to break down. This is exactly what is happening in model.ts of ec2 client. It takes a whopping 14.4GB to represent that file's AST on my local machine. This makes some sense as it's ~46,000 lines long and entirely types/namespaces.

**Solution**: We're planning to break up `models/index.ts` into multiple files.

**Comment by alexforsyth** [comment]
Yep! Working on this right now. Its not quite a simple fix but we're on it.

**Comment by trivikr** [comment]
This would soon become a blocker, as CI has started intermittently failing https://github.com/aws/aws-sdk-js-v3/pull/1461#issuecomment-680116661

**Comment by justingrant** [comment]
Thanks @trivikr. Are there any workarounds that would let me get through a local build before the root cause is fixed?  I'm in the middle of preparing a PR and would prefer to validate that it will actually build locally before trying my luck at CI. ;-)

Speaking of PRs, what I'm working on is a follow-up to my PR #393 which you closed back in Feb because its fixes were supposed to be included in another PR.  But most of the problems that #393 was fixing are still broken.  I assume I should open a new PR with the remaining fixes?

**Comment by trivikr** [comment]
> Are there any workarounds that would let me get through a local build before the root cause is fixed?

The workaround we're currently using is building on `c5.4xlarge` EC2 instances for development.
We're actively working on fixing this issue by splitting models into multiple files.

**Comment by alexforsyth** [comment]
@justingrant Thanks so much. We're currently prioritizing this internally. I generally try to keep development on my local machine (personal preference) and this workaround seems to work for me. Obviously its not ideal but its getting me unblocked in the short term.

Workaround: increase stack size allowed to node:
in file:
```
~/..../aws-sdk-js-v3/node_modules/.bin/tsc
```
make it read:
```
#!/usr/bin/env node --stack_size=2048
require('../lib/tsc.js')
```

**Comment by AllanFly120** [comment]
Hi @justingrant  I will close this issue as it has been fixed in https://github.com/aws/aws-sdk-js-v3/pull/1510. Please let me know if I need to reopen it.
